src/algebraic_geometry/projective_spectrum/structure_sheaf.lean

```diff
@@ -0,0 +1,487 @@
+      rw zero_add, },
+    { induction m using int.induction_on with m ih m ih,
+      { exact zero_mem _ },
+      { exact add_mem _ ih one_mem, },
+      { push_cast at ih ⊢,
+        exact sub_mem _ ih one_mem, } },
+  end⟩, c.denom, c.denom_not_mem⟩ }
+
+lemma deg_zsmul (c : num_denom_same_deg x) (m : ℤ) : (m • c).deg = c.deg := rfl
+lemma num_zsmul (c : num_denom_same_deg x) (m : ℤ) : ((m • c).num : A) = m • c.num := rfl
+lemma denom_zsmul (c : num_denom_same_deg x) (m : ℤ) : ((m • c).denom : A) = c.denom := rfl
+
+instance nat_scalar : has_scalar ℕ (num_denom_same_deg x) :=
+{ smul := λ m c, (m : ℤ) • c }
+
+def num_denom_same_deg.embedding (p : num_denom_same_deg x) : at x :=
+localization.mk p.num ⟨p.denom, p.denom_not_mem⟩
+
+def homogeneous_localization : Type* := quotient (setoid.ker $ num_denom_same_deg.embedding x)
+
+variable {x}
+def homogeneous_localization.val (y : homogeneous_localization x) : at x :=
+quotient.lift_on' y (num_denom_same_deg.embedding x) $ λ _ _, id
+
+variable (x)
+lemma homogeneous_localization.val_injective :
+  function.injective (@homogeneous_localization.val _ _ _ _ _ 𝒜 _ x) :=
+λ a b, quotient.rec_on_subsingleton₂' a b $ λ a b h, quotient.sound' h
+
+instance homogeneous_localization.has_pow : has_pow (homogeneous_localization x) ℕ :=
+{ pow := λ z n, @@quotient.map (setoid.ker $ num_denom_same_deg.embedding x)
+    (setoid.ker $ num_denom_same_deg.embedding x) (λ y, y^n)
+    (λ c1 c2 (h : localization.mk _ _ = localization.mk _ _), begin
+      change localization.mk _ _ = localization.mk _ _,
+      simp only [num_pow, denom_pow],
+      convert congr_arg (λ z, z ^ n) h;
+      erw localization.mk_pow;
+      refl,
+    end) z }
+
+instance : has_scalar ℤ (homogeneous_localization x) :=
+{ smul := λ m z, @@quotient.map (setoid.ker $ num_denom_same_deg.embedding x)
+    (setoid.ker $ num_denom_same_deg.embedding x) (λ y, m • y)
+    (λ c1 c2 (h : localization.mk _ _ = localization.mk _ _), begin
+      change localization.mk _ _ = localization.mk _ _,
+      simp only [num_zsmul, denom_zsmul],
+      convert congr_arg (λ z, m • z) h;
+      rw [zsmul_eq_mul, zsmul_eq_mul, show (m : at x) = localization.mk m 1, begin
+        induction m using int.induction_on with n ih n ih,
+        { erw localization.mk_zero _, refl, },
+        { push_cast,
+          erw [ih, ← localization.mk_one, localization.add_mk, one_mul, one_mul, add_comm, one_mul],
+          refl, },
+        { push_cast at ih ⊢,
+          erw [ih, ← localization.mk_one, sub_eq_add_neg, localization.neg_mk, localization.add_mk,
+            mul_one, one_mul, one_mul, add_comm, ← sub_eq_add_neg], },
+      end, localization.mk_mul, one_mul];
+      refl,
+    end) z }
+
+instance homogeneous_localization.nat_scalar : has_scalar ℕ (homogeneous_localization x) :=
+{ smul := λ n z, (n : ℤ) • z }
+
+instance : has_neg (homogeneous_localization x) :=
+{ neg := λ z, @@quotient.map (setoid.ker $ num_denom_same_deg.embedding x)
```

Comment:
Instead of using `@@`, does using `quotient.map'` and the other primed lemma names work?

Reply:
Yes, I pushed a shorter version without `@@`

Reply:
Thanks for doing this! It looks much nicer